scripts/main.js

```diff
@@ -0,0 +1,348 @@
+"use strict";
+document.addEventListener("keydown", (event) => move(event.key));
+
+const rows = document.querySelectorAll("tr");
+
+const button = document.querySelector(".button");
+const score = document.querySelector(".game-score");
+
+let scoreVal = 0;
+
+const board = [
+  [0, 0, 0, 0],
+  [0, 0, 0, 0],
+  [0, 0, 0, 0],
+  [0, 0, 0, 0],
+];
+
+const message_start = document.querySelector(".message-start");
+const message_lose = document.querySelector(".message-lose");
+const message_win = document.querySelector(".message-win");
+
+button.addEventListener("click", () => {
+  if (button.classList.contains("start")) {
+    start();
+  } else {
+    restart();
+  }
+});
+
+function start() {
+  button.innerHTML = "Restart";
+  button.classList.add("restart");
+  button.classList.remove("start");
+
+  message_start.classList.add("hidden");
+
+  newCell();
+  newCell();
+  drawBoard();
+}
+
+function restart() {
+  scoreVal = 0;
+  button.innerHTML = "Start";
+  button.classList.add("start");
+  button.classList.remove("restart");
+
+  message_win.classList.add("hidden");
+  message_lose.classList.add("hidden");
+  message_start.classList.remove("hidden");
+
+  for (let y = 0; y < 4; y++) {
+    for (let x = 0; x < 4; x++) {
+      board[y][x] = 0;
+    }
+  }
+  drawBoard();
+}
+
+function move(key) {
+  let canMove = false;
+  if (
+    message_start.classList.contains("hidden") &&
+    message_win.classList.contains("hidden") &&
+    message_lose.classList.contains("hidden")
+  ) {
+    if (key === "ArrowUp") {
+      canMove = moveUp();
+    } else if (key === "ArrowDown") {
+      canMove = moveDown();
+    } else if (key === "ArrowLeft") {
+      canMove = moveLeft();
+    } else if (key === "ArrowRight") {
+      canMove = moveRight();
+    }
+
+    if (canMove) {
+      setTimeout(() => {
+        newCell();
+        drawBoard();
+        if (board.find((row) => row.find((el) => el === 2048))) {
+          message_win.classList.remove("hidden");
+        }
+      }, 300);
+    }
+  }
+}
+
+function checkIfLose() {
+  let lose = true;
+
+  if (board.find((row) => row.includes(0))) {
+    lose = false;
+  }
+
+  for (let y = 0; y < 4; y++) {
+    for (let x = 0; x < 4; x++) {
+      if (y < 3 && board[y][x] === board[y + 1][x]) {
+        lose = false;
+      }
+
+      if (x < 3 && board[y][x] === board[y][x + 1]) {
+        lose = false;
+      }
+    }
+  }
+
+  return lose;
+}
+
+function newCell() {
+  const emptyCells = [];
+  for (let y = 0; y < 4; y++) {
+    for (let x = 0; x < 4; x++) {
+      if (!board[y][x]) {
+        emptyCells.push({ y: y, x: x });
+      }
+    }
+  }
+
+  const val = Math.floor(Math.random() * emptyCells.length);
+  board[emptyCells[val].y][emptyCells[val].x] = Math.random() < 0.9 ? 2 : 4;
+
+  if (checkIfLose()) {
+    message_lose.classList.remove("hidden");
+  }
+}
```

Comment:
The 'newCell' function not only adds a new cell to the board but also checks if the player has lost. These are separate concerns and should be handled by different functions to improve modularity and testability.